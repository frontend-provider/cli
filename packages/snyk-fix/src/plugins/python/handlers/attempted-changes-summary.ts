import {
  DependencyPins,
  FixChangesError,
  FixChangesSuccess,
  FixChangesSummary,
} from '../../../types';

export function generateFailedChanges(
  attemptedUpdates: string[],
  pins: DependencyPins,
  error: Error,
  command?: string,
): FixChangesError[] {
  const changes: FixChangesError[] = [];
  for (const pkgAtVersion of Object.keys(pins)) {
    const pin = pins[pkgAtVersion];
    if (
      !attemptedUpdates
        .map((update) => update.replace('==', '@'))
        .includes(pin.upgradeTo)
    ) {
      continue;
    }
    const updatedMessage = pin.isTransitive ? 'pin' : 'upgrade';
    const newVersion = pin.upgradeTo.split('@')[1];
    const [pkgName, version] = pkgAtVersion.split('@');

    changes.push({
      success: false,
      reason: error.message,
      userMessage: `Failed to ${updatedMessage} ${pkgName} from ${version} to ${newVersion}`,
      tip: command ? `Try running \`${command}\`` : undefined,
      issueIds: pin.vulns,
      from: pkgAtVersion,
      to: `${pkgName}@${newVersion}`,
    });
  }
  return changes;
}

export function generateSuccessfulChanges(
  appliedUpgrades: string[],
  pins: DependencyPins,
): FixChangesSuccess[] {
  const changes: FixChangesSuccess[] = [];
  for (const pkgAtVersion of Object.keys(pins)) {
    const pin = pins[pkgAtVersion];
    if (
      !appliedUpgrades
        .map((upgrade) => upgrade.replace('==', '@'))
        .includes(pin.upgradeTo)
    ) {
      continue;
    }
    const updatedMessage = pin.isTransitive ? 'Pinned' : 'Upgraded';
    const newVersion = pin.upgradeTo.split('@')[1];
    const [pkgName, version] = pkgAtVersion.split('@');

    changes.push({
      success: true,
      userMessage: `${updatedMessage} ${pkgName} from ${version} to ${newVersion}`,
      issueIds: pin.vulns,
      from: pkgAtVersion,
      to: `${pkgName}@${newVersion}`,
    });
  }
  return changes;
}

export function isSuccessfulChange(
  change: FixChangesSummary,
): change is FixChangesError {
  return change.success === true;
}
