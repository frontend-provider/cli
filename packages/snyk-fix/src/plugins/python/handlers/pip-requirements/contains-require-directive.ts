/* Requires like -r, -c are not supported at the moment, as multiple files
 * would have to be identified and fixed together
 * https://pip.pypa.io/en/stable/reference/pip_install/#options
 */
export async function containsRequireDirective(
  requirementsTxt: string,
): Promise<{ containsRequire: boolean; matches: RegExpMatchArray[] }> {
  const allMatches: RegExpMatchArray[] = [];
  const REQUIRE_PATTERN = new RegExp(/^[^\S\n]*-(r|c)\s+(.+)/, 'gm');
  const matches = getAllMatchedGroups(REQUIRE_PATTERN, requirementsTxt);
  for (const match of matches) {
    if (match && match.length > 1) {
      allMatches.push(match);
    }
  }
  return { containsRequire: allMatches.length > 0, matches: allMatches };
}

function getAllMatchedGroups(re: RegExp, str: string) {
  const groups: RegExpExecArray[] = [];
  let match: RegExpExecArray | null;
  while ((match = re.exec(str))) {
    groups.push(match);
  }

  return groups;
}
