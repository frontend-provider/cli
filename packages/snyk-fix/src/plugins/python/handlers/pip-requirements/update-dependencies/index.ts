import * as debugLib from 'debug';

import { NoFixesCouldBeAppliedError } from '../../../../../lib/errors/no-fixes-applied';
import { DependencyPins, FixChangesSummary } from '../../../../../types';
import { generatePins } from './generate-pins';
import { applyUpgrades } from './apply-upgrades';
import { parseRequirementsFile } from './requirements-file-parser';
import { generateUpgrades } from './generate-upgrades';
import { FailedToParseManifest } from '../../../../../lib/errors/failed-to-parse-manifest';

const debug = debugLib('snyk-fix:python:update-dependencies');

/*
 * Given contents of manifest file(s) and a set of upgrades, apply the given
 * upgrades to a manifest and return the upgraded manifest.
 *
 * Currently only supported for `requirements.txt` - at least one file named
 * `requirements.txt` must be in the manifests.
 */
export function updateDependencies(
  requirementsTxt: string,
  updates: DependencyPins,
): { updatedManifest: string; changes: FixChangesSummary[] } {
  const parsedRequirementsData = parseRequirementsFile(requirementsTxt);

  if (!parsedRequirementsData.length) {
    debug(
      'Error: Expected to receive parsed manifest data. Is manifest empty?',
    );
    throw new FailedToParseManifest();
  }
  debug('Finished parsing manifest');

  const { updatedRequirements, changes: upgradedChanges } = generateUpgrades(
    parsedRequirementsData,
    updates,
  );
  debug('Finished generating upgrades to apply');

  const { pinnedRequirements, changes: pinChanges } = generatePins(
    parsedRequirementsData,
    updates,
  );
  debug('Finished generating pins to apply');

  let updatedManifest = [
    ...applyUpgrades(parsedRequirementsData, updatedRequirements),
    ...pinnedRequirements,
  ].join('\n');

  // This is a bit of a hack, but an easy one to follow. If a file ends with a
  // new line, ensure we keep it this way. Don't hijack customers formatting.
  if (requirementsTxt.endsWith('\n')) {
    updatedManifest += '\n';
  }
  debug('Finished applying changes to manifest');

  // TODO: do this with the changes now that we only return new
  if (updatedManifest === requirementsTxt) {
    debug('Manifest has not changed!');
    throw new NoFixesCouldBeAppliedError();
  }

  return {
    updatedManifest,
    changes: [...pinChanges, ...upgradedChanges],
  };
}
