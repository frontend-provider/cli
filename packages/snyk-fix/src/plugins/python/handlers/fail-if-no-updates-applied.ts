import * as debugLib from 'debug';

import { NoFixesCouldBeAppliedError } from '../../../lib/errors/no-fixes-applied';
import { FixChangesError, FixChangesSummary } from '../../../types';
import { isSuccessfulChange } from './attempted-changes-summary';

const debug = debugLib('snyk-fix:python:ensure-changes-applied');

export function failIfNoUpdatesApplied(changes: FixChangesSummary[]) {
  if (!changes.length) {
    throw new NoFixesCouldBeAppliedError();
  }
  if (!changes.some((c) => isSuccessfulChange(c))) {
    debug('Manifest has not changed as no changes got applied!');
    // throw the first error tip since 100% failed, they all failed with the same
    // error
    const { reason, tip } = changes[0] as FixChangesError;
    throw new NoFixesCouldBeAppliedError(reason, tip);
  }
}
