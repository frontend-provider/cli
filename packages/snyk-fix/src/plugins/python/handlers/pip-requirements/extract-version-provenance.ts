import * as path from 'path';
import * as debugLib from 'debug';

import {
  ParsedRequirements,
  parseRequirementsFile,
} from './update-dependencies/requirements-file-parser';
import { Workspace } from '../../../../types';
import { containsRequireDirective } from './contains-require-directive';

export interface PythonProvenance {
  [fileName: string]: ParsedRequirements;
}

const debug = debugLib('snyk-fix:python:extract-version-provenance');

export async function extractProvenance(
  workspace: Workspace,
  dir: string,
  fileName: string,
  provenance: PythonProvenance = {},
): Promise<PythonProvenance> {
  const requirementsFileName = path.join(dir, fileName);
  const requirementsTxt = await workspace.readFile(requirementsFileName);
  provenance = {
    ...provenance,
    [fileName]: parseRequirementsFile(requirementsTxt),
  };
  const { containsRequire, matches } = await containsRequireDirective(
    requirementsTxt,
  );
  if (containsRequire) {
    for (const match of matches) {
      const requiredFilePath = match[2];
      if (provenance[requiredFilePath]) {
        debug('Detected recursive require directive, skipping');
        continue;
      }

      provenance = {
        ...provenance,
        ...(await extractProvenance(
          workspace,
          dir,
          requiredFilePath,
          provenance,
        )),
      };
    }
  }
  return provenance;
}
