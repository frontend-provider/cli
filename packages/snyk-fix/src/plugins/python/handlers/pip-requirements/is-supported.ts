import * as chalk from 'chalk';

import { EntityToFix, WithUserMessage } from '../../../../types';
import { containsRequireDirective } from './contains-require-directive';

interface Supported {
  supported: true;
}

interface NotSupported {
  supported: false;
  reason: string;
}

export function projectTypeSupported(
  res: Supported | NotSupported,
): res is Supported {
  return !('reason' in res);
}

export async function isSupported(
  entity: EntityToFix,
): Promise<Supported | NotSupported> {
  const remediationData = entity.testResult.remediation;

  if (!remediationData) {
    return { supported: false, reason: 'No remediation data available' };
  }
  if (!remediationData.pin || Object.keys(remediationData.pin).length === 0) {
    return {
      supported: false,
      reason: 'There is no actionable remediation to apply',
    };
  }

  // TODO: fix the non null assertion here
  let requirementsTxt;
  try {
    const fileName = entity.scanResult.identity.targetFile!;
    requirementsTxt = await entity.workspace.readFile(fileName);
  } catch (e) {
    return {
      supported: false,
      reason: e.message,
    };
  }

  const { containsRequire } = await containsRequireDirective(requirementsTxt);
  if (containsRequire) {
    return {
      supported: false,
      reason: `Requirements with ${chalk.bold('-r')} or ${chalk.bold(
        '-c',
      )} directive are not yet supported`,
    };
  }

  return { supported: true };
}

export async function partitionByFixable(
  entities: EntityToFix[],
): Promise<{
  skipped: Array<WithUserMessage<EntityToFix>>;
  fixable: EntityToFix[];
}> {
  const fixable: EntityToFix[] = [];
  const skipped: Array<WithUserMessage<EntityToFix>> = [];
  for (const entity of entities) {
    const isSupportedResponse = await isSupported(entity);
    if (projectTypeSupported(isSupportedResponse)) {
      fixable.push(entity);
      continue;
    }
    skipped.push({
      original: entity,
      userMessage: isSupportedResponse.reason,
    });
  }
  return { fixable, skipped };
}
