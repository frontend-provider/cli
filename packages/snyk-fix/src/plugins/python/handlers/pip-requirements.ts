import * as debugLib from 'debug';
import * as chalk from 'chalk';

import { EntityToFix, FixOptions, WithFixChangesApplied } from '../../../types';
import { PluginFixResponse } from '../../types';
import { updateDependencies } from './update-dependencies';
import { MissingRemediationDataError } from '../../../lib/errors/missing-remediation-data';
import { MissingFileNameError } from '../../../lib/errors/missing-file-name';

const debug = debugLib('snyk-fix:python:requirements.txt');

export async function pipRequirementsTxt(
  entities: EntityToFix[],
  options: FixOptions,
): Promise<PluginFixResponse> {
  debug(`Preparing to fix ${entities.length} Python requirements.txt projects`);
  const handlerResult: PluginFixResponse = {
    succeeded: [],
    failed: [],
    skipped: [],
  };
  for (const entity of entities) {
    try {
      const isSupportedResponse = await isSupported(entity);
      if (projectTypeSupported(isSupportedResponse)) {
        const fixedEntity = await fixIndividualRequirementsTxt(entity, options);
        handlerResult.succeeded.push(fixedEntity);
      } else {
        handlerResult.skipped.push({
          original: entity,
          userMessage: isSupportedResponse.reason,
        });
      }
    } catch (e) {
      handlerResult.failed.push({ original: entity, error: e });
    }
  }
  return handlerResult;
}

function projectTypeSupported(res: Supported | NotSupported): res is Supported {
  return !('reason' in res);
}

interface Supported {
  supported: true;
}

interface NotSupported {
  supported: false;
  reason: string;
}
export async function isSupported(
  entity: EntityToFix,
): Promise<Supported | NotSupported> {
  const remediationData = entity.testResult.remediation;
  if (!remediationData) {
    return { supported: false, reason: 'No remediation data available' };
  }
  // TODO: recursive inclusions?
  // TODO: fix the non null assertion here
  const fileName = entity.scanResult.identity.targetFile!;
  const requirementsTxt = await entity.workspace.readFile(fileName);
  const { containsRequire } = await containsRequireDirective(requirementsTxt);

  if (containsRequire) {
    return {
      supported: false,
      reason: `Requirements with ${chalk.bold('-r')} or ${chalk.bold(
        '-c',
      )} directive are not yet supported`,
    };
  }

  if (!remediationData.pin || Object.keys(remediationData.pin).length === 0) {
    return {
      supported: false,
      reason: 'There is no actionable remediation to apply',
    };
  }
  return { supported: true };
}

/* Requires like -r, -c are not supported at the moment, as multiple files
 * would have to be identified and fixed together
 * https://pip.pypa.io/en/stable/reference/pip_install/#options
 */
export async function containsRequireDirective(
  requirementsTxt: string,
): Promise<{ containsRequire: boolean; matches: RegExpMatchArray[] }> {
  const allMatches: RegExpMatchArray[] = [];
  const REQUIRE_PATTERN = new RegExp(/^[^\S\n]*-(r|c)\s+(.+)/, 'gm');
  const matches = requirementsTxt.matchAll(REQUIRE_PATTERN);
  for (const match of matches) {
    if (match && match.length > 1) {
      allMatches.push(match);
    }
  }
  return { containsRequire: allMatches.length > 0, matches: allMatches };
}

// TODO: optionally verify the deps install
export async function fixIndividualRequirementsTxt(
  entity: EntityToFix,
  options: FixOptions,
): Promise<WithFixChangesApplied<EntityToFix>> {
  const fileName = entity.scanResult.identity.targetFile;
  const remediationData = entity.testResult.remediation;
  if (!remediationData) {
    throw new MissingRemediationDataError();
  }
  if (!fileName) {
    throw new MissingFileNameError();
  }
  const requirementsTxt = await entity.workspace.readFile(fileName);
  // TODO: allow handlers per fix type (later also strategies or combine with strategies)
  const { updatedManifest, changes } = updateDependencies(
    requirementsTxt,
    remediationData.pin,
  );
  if (!options.dryRun) {
    await entity.workspace.writeFile(fileName, updatedManifest);
  }

  return {
    original: entity,
    changes,
  };
}
