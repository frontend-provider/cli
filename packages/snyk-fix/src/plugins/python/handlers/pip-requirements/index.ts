import * as debugLib from 'debug';
import * as pathLib from 'path';

import {
  DependencyPins,
  EntityToFix,
  FixChangesSummary,
  FixOptions,
  RemediationChanges,
  Workspace,
} from '../../../../types';
import { PluginFixResponse } from '../../../types';
import { updateDependencies } from './update-dependencies';
import { MissingRemediationDataError } from '../../../../lib/errors/missing-remediation-data';
import { MissingFileNameError } from '../../../../lib/errors/missing-file-name';
import { partitionByFixable } from './is-supported';
import { NoFixesCouldBeAppliedError } from '../../../../lib/errors/no-fixes-applied';
import {
  extractProvenance,
  PythonProvenance,
} from './extract-version-provenance';
import {
  ParsedRequirements,
  parseRequirementsFile,
  Requirement,
} from './update-dependencies/requirements-file-parser';

const debug = debugLib('snyk-fix:python:requirements.txt');

export async function pipRequirementsTxt(
  entities: EntityToFix[],
  options: FixOptions,
): Promise<PluginFixResponse> {
  debug(`Preparing to fix ${entities.length} Python requirements.txt projects`);
  const handlerResult: PluginFixResponse = {
    succeeded: [],
    failed: [],
    skipped: [],
  };

  const { fixable, skipped } = await partitionByFixable(entities);
  handlerResult.skipped.push(...skipped);

  for (const entity of fixable) {
    try {
      const { changes } = await applyAllFixes(
        entity,
        // dir,
        // base,
        // remediation,
        // provenance,
        options,
      );
      if (!changes.length) {
        debug('Manifest has not changed!');
        throw new NoFixesCouldBeAppliedError();
      }
      handlerResult.succeeded.push({ original: entity, changes });
    } catch (e) {
      handlerResult.failed.push({ original: entity, error: e });
    }
  }
  return handlerResult;
}

export function getRequiredData(
  entity: EntityToFix,
): {
  remediation: RemediationChanges;
  targetFile: string;
  workspace: Workspace;
} {
  const { remediation } = entity.testResult;
  if (!remediation) {
    throw new MissingRemediationDataError();
  }
  const { targetFile } = entity.scanResult.identity;
  if (!targetFile) {
    throw new MissingFileNameError();
  }
  const { workspace } = entity;
  if (!workspace) {
    throw new NoFixesCouldBeAppliedError();
  }
  return { targetFile, remediation, workspace };
}

// TODO: optionally verify the deps install
export async function fixIndividualRequirementsTxt(
  workspace: Workspace,
  dir: string,
  entryFileName: string,
  fileName: string,
  remediation: RemediationChanges,
  parsedRequirements: ParsedRequirements,
  options: FixOptions,
  directUpgradesOnly: boolean,
): Promise<{ changes: FixChangesSummary[]; appliedRemediation: string[] }> {
  const fullFilePath = pathLib.join(dir, fileName);
  const { updatedManifest, changes, appliedRemediation } = updateDependencies(
    parsedRequirements,
    remediation.pin,
    directUpgradesOnly,
    pathLib.join(dir, entryFileName) !== fullFilePath ? fileName : undefined,
  );
  if (!options.dryRun && changes.length > 0) {
    debug('Writing changes to file');
    await workspace.writeFile(pathLib.join(dir, fileName), updatedManifest);
  } else {
    debug('Skipping writing changes to file in --dry-run mode');
  }

  return { changes, appliedRemediation };
}

export async function applyAllFixes(
  entity: EntityToFix,
  options: FixOptions,
): Promise<{ changes: FixChangesSummary[] }> {
  const { remediation, targetFile: entryFileName, workspace } = getRequiredData(
    entity,
  );
  const { dir, base } = pathLib.parse(entryFileName);
  const provenance = await extractProvenance(workspace, dir, base);
  const upgradeChanges: FixChangesSummary[] = [];
  const appliedUpgradeRemediation: string[] = [];
  for (const fileName of Object.keys(provenance)) {
    const skipApplyingPins = true;
    const { changes, appliedRemediation } = await fixIndividualRequirementsTxt(
      workspace,
      dir,
      base,
      fileName,
      remediation,
      provenance[fileName],
      options,
      skipApplyingPins,
    );
    appliedUpgradeRemediation.push(...appliedRemediation);
    // what if we saw the file before and already fixed it?
    upgradeChanges.push(...changes);
  }
  // now do left overs as pins + add tests
  const requirementsTxt = await workspace.readFile(entryFileName);

  const toPin: RemediationChanges = filterOutAppliedUpgrades(
    remediation,
    appliedUpgradeRemediation,
  );
  const directUpgradesOnly = false;
  const { changes: pinnedChanges } = await fixIndividualRequirementsTxt(
    workspace,
    dir,
    base,
    base,
    toPin,
    parseRequirementsFile(requirementsTxt),
    options,
    directUpgradesOnly,
  );

  return { changes: [...upgradeChanges, ...pinnedChanges] };
}

function filterOutAppliedUpgrades(
  remediation: RemediationChanges,
  appliedRemediation: string[],
): RemediationChanges {
  const pinRemediation: RemediationChanges = {
    ...remediation,
    pin: {}, // delete the pin remediation so we can add only not applied
  };
  const pins = remediation.pin;
  const lowerCasedAppliedRemediation = appliedRemediation.map((i) =>
    i.toLowerCase(),
  );
  for (const pkgAtVersion of Object.keys(pins)) {
    if (!lowerCasedAppliedRemediation.includes(pkgAtVersion.toLowerCase())) {
      pinRemediation.pin[pkgAtVersion] = pins[pkgAtVersion];
    }
  }
  return pinRemediation;
}
