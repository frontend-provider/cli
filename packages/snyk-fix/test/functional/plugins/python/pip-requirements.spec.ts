import { isSupported } from '../../../../src/plugins/python/handlers/pip-requirements';
import { generateEntityToFix } from '../../../helpers/generate-entity-to-fix';

describe('isSupported', () => {
  it('with missing remediation is not supported', async () => {
    const entity = generateEntityToFix(
      'pip',
      'requirements.txt',
      JSON.stringify({}),
    );
    delete entity.testResult.remediation;
    expect(await isSupported(entity)).toBeFalsy();
  });
  it('with -r directive in the manifest not supported', async () => {
    const entity = generateEntityToFix(
      'pip',
      'requirements.txt',
      '-r prod.txt\nDjango==1.6.1',
    );
    expect(await isSupported(entity)).toBeFalsy();
  });
  it('with -c directive in the manifest not supported', async () => {
    const entity = generateEntityToFix(
      'pip',
      'requirements.txt',
      '-c constraints.txt',
    );
    expect(await isSupported(entity)).toBeFalsy();
  });
  it('with -e directive in the manifest is supported', async () => {
    const entity = generateEntityToFix('pip', 'requirements.txt', '-e .');
    expect(await isSupported(entity)).toBeTruthy();
  });
});
