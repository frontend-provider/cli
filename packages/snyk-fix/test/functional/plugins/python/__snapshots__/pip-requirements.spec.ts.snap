// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`containsRequireDirective with -c 1`] = `
Object {
  "containsRequire": true,
  "matches": Array [
    Array [
      "-c constraints.txt",
      "c",
      "constraints.txt",
    ],
  ],
}
`;

exports[`containsRequireDirective with -r & -c 1`] = `
Object {
  "containsRequire": true,
  "matches": Array [
    Array [
      "-c constraints.txt",
      "c",
      "constraints.txt",
    ],
    Array [
      "-r base.txt",
      "r",
      "base.txt",
    ],
  ],
}
`;

exports[`containsRequireDirective with -r & -c after deps 1`] = `
Object {
  "containsRequire": true,
  "matches": Array [
    Array [
      "-c constraints.txt",
      "c",
      "constraints.txt",
    ],
    Array [
      "-r base.txt",
      "r",
      "base.txt",
    ],
  ],
}
`;

exports[`containsRequireDirective with -r 1`] = `
Object {
  "containsRequire": true,
  "matches": Array [
    Array [
      "-r requirements.txt",
      "r",
      "requirements.txt",
    ],
  ],
}
`;

exports[`containsRequireDirective with multiple -r 1`] = `
Object {
  "containsRequire": true,
  "matches": Array [
    Array [
      "-r requirements.txt",
      "r",
      "requirements.txt",
    ],
    Array [
      "-r base.txt",
      "r",
      "base.txt",
    ],
  ],
}
`;
